src/librustc/session/mod.rs

```diff
@@ -71,7 +73,58 @@ impl Session {
         self.diagnostic().handler().fatal(msg)
     }
     pub fn span_err(&self, sp: Span, msg: &str) {
-        self.diagnostic().span_err(sp, msg)
+        // Conditions for enabling multi-line errors:
+        if !msg.contains("mismatched types") &&
+           !msg.contains("type mismatch resolving") &&
+           !msg.contains("if and else have incompatible types") &&
+           !msg.contains("if may be missing an else clause") &&
+           !msg.contains("match arms have incompatible types") &&
+           !msg.contains("structure constructor specifies a structure of type") {
+            return self.diagnostic().span_err(sp, msg);
+        }
+
+        let first  = Regex::new(r"[( ]expected").unwrap();
+        let second = Regex::new(r" found").unwrap();
+        let third  = Regex::new(
+            r"\((values differ|lifetime|cyclic type of infinite size)").unwrap();
```

Comment:
You want this it trimmed? Raw strings have no escapes and breaking them can give surprising results when applied with a regex. Using non-raw strings also can give surprising results.

Technically I could do this but I decided against it because it makes it much less obvious what you're trying to do.
```rust
let s = r"\((values differ|lifetime|cyclic type of infinite size)";
// could also be
let v = vec![r"\((values differ",
             r"lifetime",
             r"cyclic type of infinite size)"].connect("|");
```